docs/smt/Cardinality.md

```diff
@@ -0,0 +1,209 @@
+# SMT encoding for set cardinalities
+
+**Author: Igor Konnov, March 2020**
+
+This document contains rationale and the proposal for a better SMT encoding of
+the `Cardinality` operator. The encoding of v0.6.0 is explained in the
+[OOPSLA19 paper](https://dl.acm.org/doi/10.1145/3360549).  In the following, we
+are using the terminology of that paper. Moreover, we use the TLA+ ASCII
+notation, wherever possible.
+
+## 1. Problem
+
+Assume that a set `S` is over-approximated with arena cells `c_1, ..., c_n`.
+Some of these cells actually belong to the set, and some do not. A Boolean
+constant `in_i`, if and only if cell `c_i` belongs to the set `S`.
+
+In the following, we assume that the equality `=` is interpreted on `c_1`, ...,
+`c_n` as the structural equality. As Apalache is using lazy equality, we assume
+that the equality constraints `c_i = c_j` have been constructed for `i, j \in
+1..n`.
+
+## 2. The quadratic encoding in v0.5.0 and v0.6.0
+
+In this encoding, we introduce `n + 1` counters `k_0`, `k_1`, ..., `k_n`. The
+counter `k_i` contains the cardinality of the set that is over-approximated
+with the cells `c_1`, ..., `c_i`. The counters are defined by the following
+equations for `i \in 1..n`.
+
+```tla
+k_0 = 0
+k_i =  k_{i-1} + (IF in_{i+1} /\ notSeen_i THEN 1 ELSE 0)
+```
+
+where the formula `notSeen_i` is defined as follows:
+
+```tla
+notSeen_i == \A j \in 1..(i-1): ~in_j \/ c_j /= c_i
+```
+
+**Proposition 1.** `Cardinality(S) = k_n`.
+
+**Complexity.** To get an idea of complexity, we count the number of literals
+that are produced by this encoding. It is easy to see that every equation for
+`k_i` produces `O(i)` literals, `i` inequality tests appear in the formula
+`notSeen_i`. If we compute the sum 0 + ... + n - 1, we get `(n - 1) * n`.
+Hence, our encoding produces `O(n^2)` literals, that is, it is **quadratic** in
+the number of elements that over-approximate `S`.
+
+## 3. A linear encoding for a comparison against a constant in v0.7.0
+
+**Lower bounds.** We have developed a more efficient encoding for the formulas
+that have the following form:
+
+```tla
+Cardinality(S) >= k,
+```
+
+where `S` is a set, and `k` is a constant integer expression.
+
+Essentially, this encoding can be translated into the following formula
+`[ExistsForm]`:
+
+```tla
+\E x_1, ..., x_k \in S:
+  \A y, z \in {x_1, ..., x_k}:
+    y /= z
+```
+
```

Comment:
I would add here an explanation about the complexity of this encoding. E.g., as far as I understand, it is linear in the number of set elements, but quadratic in the constant, is it true? Which also raises an interesting question on how this encoding will behave when we compare a small set to a large constant, and whether actually we should choose either one encoding or another depending on the relationship between the number of set cells and the constant.

Reply:
To be precise, the encoding is linear in the length of the set approximation, which tends to grow faster than the set itself (e.g., `S' = S \ {x}` makes the set `S` smaller, but its approximation has the same size).